### ISSUE TITLE
the generated faces are not real when converting 3d face meshes to real faces

### ISSUE THREAD

**Issue by Honlan** [eladrich/pixel2style2pixel#174]
Hi~

Thanks a lot for your awesome work.
I am trying to translate 3d face meshes to real faces based on the FFHQ dataset by using the following command:

`python scripts/train.py \
--dataset_type=ffhq_3d_to_face \
--exp_dir=exp/ffhq_3d_to_face/ \
--workers=8 \
--batch_size=8 \
--test_batch_size=8 \
--test_workers=8 \
--val_interval=2500 \
--save_interval=5000 \
--encoder_type=GradualStyleEncoder \
--start_from_latent_avg \
--lpips_lambda=0.8 \
--l2_lambda=1 \
--id_lambda=0.1 \
--w_norm_lambda=0.005 \
--output_size=256`

where the 3d face meshes are obtained by inference on FFHQ with 3DDFA_v2(https://github.com/cleardusk/3DDFA_V2). After training for 50W iterations, however, the generated faces are not as real as those of the paper.

Could you please give me some possible advices? Thanks again.

![图片 1](https://user-images.githubusercontent.com/5653648/126263868-6c9d541b-b0bd-4cea-903e-f0d85dcc60d0.png)

**Comment by yuval-alaluf** [comment]
Hi @Honlan ,
The application is quite interesting and it's nice to the use of the 3D meshes as input to pSp. 
Are the two results you showed here on the train? Its surprising that the model is able to almost perfectly reconstruct the individual's hair style even though it is not part of the input. 
Do you have any results on the test set?

**Comment by Honlan** [comment]
The above image is from the training set. I didn't specially prepare another image set and simply use FFHQ namely the training set for test (_test_source_root_ and _test_target_root_). What a pity :(

I found that in the paper, only ffhq_encode and ffhq_frontalize are trained on FFHQ, while celebs_sketch_to_face and celebs_seg_to_face, where the input and output have larger domain gaps, are trained on CelebAMask-HQ.

As a result, may the above unrealism caused by the diversity of FFHQ?

**Comment by yuval-alaluf** [comment]
Good question. I don't think that FFHQ is the problem here. I'd say that this task reminds me a lot of the sketch to image to segmentation to image tasks we explored with pSp. And in those tasks we incorporated two changes with respect to the standard encoding task: 
1. No identity loss: after all, you have a one to many mapping here where a single 3d mesh can correspond to many possible images. 
2. Added a small w-regularization loss: this helped reduce some artifacts in these ambiguous tasks. 

So I would recommend trying to apply the loss scheme we used in the sketch/seg tasks, which are more aligned with what you're trying to do. Hopefully you'll see some more realistic results. 

Also remember that your task is a classic example where our style mixing idea can really be useful. You'll be able to easily generate an endless collection of images for each input mesh :)

**Comment by Honlan** [comment]
@yuval-alaluf thanks a lot for you advices. I will have a try :)

**Comment by Honlan** [comment]
the problem seems to be caused by `--output_size=256`.

If I add the above option, after training 1 iteration and global_step is still 0, the saved image logs/images/train/faces/0000.jpg looks like this, which is blurred and unrealistic.

![3d_0000](https://user-images.githubusercontent.com/5653648/126465690-a0232775-2691-4778-9de4-870e2bccba65.jpg)


However, if I don't add the option so output_size=1024 by default, the save image logs/images/train/faces/0000.jpg looks like this, which is clear and realistic.

![0000](https://user-images.githubusercontent.com/5653648/126465892-0583d862-da0a-4a1e-b6bf-ab673247c1de.jpg)

I further try the seg_to_face experiment, where the `--output_size` is not specified as well, and get this, which also looks reasonable.

![seg_0000](https://user-images.githubusercontent.com/5653648/126466050-92783872-8430-4847-a06a-eba500f29313.jpg)

I don't know why  output_size=256 can result in such a large difference. Do you have some possible explanation for that? Thanks.

**Comment by yuval-alaluf** [comment]
Does your generator output a resolution of 1024x1024? If so, then maybe specifying an output size that is amall from what the generator outputs can cause this.

**Comment by Honlan** [comment]
@yuval-alaluf I think I misunderstood the meaning of output_size, which should be exactly the same as the output resolution of the generator. "Added supported for StyleGANs of different resolutions (e.g., 256, 512, 1024). This can be set using the flag --output_size, which is set to 1024 by default".

I thought that for the same generator with 1024 output resolution, output_size=1024 will output 1024 image, output_size=512 will output 512 image, and output_size=256 will output 256 image. So I set output_size=256, which I thought could help save training time. Now I get the right meaning of output_size.

Thanks again for your helps~

**Comment by Honlan** [comment]
After fixing the aforementioned problem, I can obtain the expected results now.

For the training set, the reconstructed images can well preserve the original identity. However, the identities shift a lot when testing (I use 69900 images of FFHQ for training and the left 100 for testing).

Do you have any suggestions on better preserving the original identities for testing?

Here is my command for training the model, where w_id is already applied.

python scripts/train.py \
--dataset_type=ffhq_3d_to_face \
--exp_dir=exp/ffhq_3d_to_face_w_id/ \
--workers=8 \
--batch_size=8 \
--test_batch_size=8 \
--test_workers=8 \
--val_interval=2500 \
--save_interval=5000 \
--encoder_type=GradualStyleEncoder \
--start_from_latent_avg \
--lpips_lambda=0.8 \
--l2_lambda=1 \
--id_lambda=0.1 \
--w_norm_lambda=0.005

Thanks a lot :)

**Comment by Honlan** [comment]
thanks a lot for your helpful suggestions. I do this project just due to my personal interests. :)

I plan to try some augmentations and will post the results here if I have any progresses.

thanks again. :)

**Comment by ry85** [comment]
@yuval-alaluf In the stylegan generator what if we feed some frontal facing reference image of the person (let it be X_ref) to the top few layers (since they control coarse features) and the 3d mesh info to the later layers .. can we expect the stylegan generator to produce the output image that looks like X_ref and have pose from the 3d mesh.

Also I have one doubt since the trained stylegan model that is provided with this repo is of 1024 output resolution, that means when using this repo output_size should be equal to 1024 only.
